main.tf

```diff
@@ -1,65 +1,106 @@
+  for_each = { for cidr in var.vpc_ipam.pool_cidrs : cidr.name => cidr }
+
+  cidr = each.value.cidr
+  dynamic "cidr_authorization_context" {
+    for_each = each.value.cidr_authorization_context != null ? [1] : []
+
+    content {
+      message   = each.value.cidr_authorization_context_message
+      signature = each.value.cidr_authorization_context_signature
+    }
+  }
+
+  #pools could be one or more, might need to make a k,v map of CIDR foreach, then reference it's index to accurately identify
+  #compare ipam_pool_id input to pools.name, and code that way.
+  #or describe in help file, to only create ONE pool at a time.
+
+  ipam_pool_id   = aws_vpc_ipam_pool.this[var.vpc_ipam.pools[0].name].id
+  netmask_length = each.value.netmask_length
+
+  lifecycle {
+    precondition {
+      condition     = each.value.cidr != "" && each.value.netmask_length != ""
+      error_message = "Cannot set both 'cidr' and 'netmask_length'."
+    }
+  }
 }
 
-module "vpc_ipam_resource_discovery_association" {
-  source                     = "./modules/resource-discovery-association"
-  ipam_id                    = var.resource_discovery_association.ipam_id
-  ipam_resource_discovery_id = var.resource_discovery_association.ipam_resource_discovery_id
-  tags                       = var.resource_discovery_association.tags
+resource "aws_vpc_ipam_pool_cidr_allocation" "this" {
+  for_each = { for allocation in var.vpc_ipam.pool_cidr_allocations : allocation.cidr => allocation }
+
+  cidr             = each.value.cidr
+  description      = each.value.description
+  disallowed_cidrs = each.value.disallowed_cidrs
+  ipam_pool_id     = each.value.ipam_pool_id
+  netmask_length   = each.value.netmask_length
 }
 
-module "vpc_ipam_preview_next_cidr" {
-  source           = "./modules/preview-next-cidr"
-  disallowed_cidrs = var.preview_next_cidr.disallowed_cidrs
-  ipam_pool_id     = var.preview_next_cidr.ipam_pool_id
-  netmask_length   = var.preview_next_cidr.netmask_length
+# unsure of implementation usage? could be useful for checking/assigning next CIDR (dynamically)?
+resource "aws_vpc_ipam_preview_next_cidr" "this" {
+  count = var.vpc_ipam.preview_next_cidr != null ? 1 : 0
+
+  disallowed_cidrs = var.vpc_ipam.preview_next_cidr.disallowed_cidrs
+  ipam_pool_id     = var.vpc_ipam.preview_next_cidr.ipam_pool_id
+  netmask_length   = var.vpc_ipam.preview_next_cidr.netmask_length
 }
 
-module "vpc_ipam_scope" {
-  source  = "./modules/scope"
-  ipam_id = var.scope.ipam_id
-  tags    = var.scope.tags
+resource "aws_vpc_ipam_scope" "this" {
+  for_each = { for scope in var.vpc_ipam.scopes : scope.name => scope }
+
+  ipam_id     = aws_vpc_ipam.this[0].id
+  description = each.value.description
+  tags        = each.value.tags
+
+  # lifecycle {
+  #   precondition {
+  #     condition     = var.vpc_ipam.scopes[0].name != "" && var.vpc_ipam.ipam.name == ""
+  #     error_message = "You can't create a scope without creating an IPAM"
+  #   }
+  # }
```

Comment:
this lifecycle is not working as a precondition. Something is going on with the type conversion. i was explicit down to a string type, but it keeps evaluating to false.

`╷
│ Error: Resource precondition failed
│ 
│   on ../../main.tf line 102, in resource "aws_vpc_ipam_scope" "this":
│  102:       condition     = var.vpc_ipam.scopes[0].name != "" && var.vpc_ipam.ipam.name == ""
│     ├────────────────
│     │ var.vpc_ipam.ipam.name is "default_ipam"
│     │ var.vpc_ipam.scopes[0].name is "high_level_container"
│ 
│ You can't create a scope without creating an IPAM`

Reply:
This should work if the precondition is only evaluated when `scopes` is not empty. If it is not empty, then `name` is required, and we do not need to test it.